src/datatype/memory_veci8.rs

```diff
@@ -0,0 +1,241 @@
+/// We can transform F32 to I8 using the formula (a - (max + min) / 2) / (max - min) * 255, resulting in a vector of I8, [b_0, b_1, b_2, ..., b_n]. Here 255 is the range size that the int8 type can cover, which is the difference between -128 and 127.
+/// Converting I8 back to F32 can be achieved by using the formula b * (max - min) / 255 + (max + min) / 2, which gives us a vector of F32, albeit with a slight loss of precision.
+/// We use alpha to represent (max - min) / 255, and offset to represent (max + min) / 2 here.
+#[repr(C, align(8))]
+pub struct Veci8Header {
+    varlena: u32,
+    len: u16,
+    kind: u8,
+    reserved: u8,
+    alpha: F32,
+    offset: F32,
+    phantom: [I8; 0],
+}
+
+impl Veci8Header {
+    fn varlena(size: usize) -> u32 {
+        // varattrib_4b type with compress is not set
+        (size << 2) as u32
+    }
+
+    fn layout(len: usize) -> Layout {
+        u16::try_from(len).expect("Vector is too large.");
+        // size of struct VecI8
+        let layout_alpha = Layout::new::<Veci8Header>();
+        // size of data in VecI8
+        let layout_beta = Layout::array::<I8>(len).unwrap();
+        let layout = layout_alpha.extend(layout_beta).unwrap().0;
+        layout.pad_to_align()
+    }
+
+    pub fn len(&self) -> usize {
+        self.len as usize
+    }
+
+    pub fn alpha(&self) -> F32 {
+        self.alpha
+    }
+
+    pub fn offset(&self) -> F32 {
+        self.offset
+    }
+
+    /// return value after dequantization by index
+    /// since index<usize> return &Output, we can't create a new Output and return it as a reference, so we need to use this function to return a new Output directly
+    #[inline(always)]
+    pub fn index(&self, index: usize) -> F32 {
+        self.data()[index].to_f32() * self.alpha() + self.offset()
+    }
+
+    pub fn to_ref(&self) -> Veci8Borrowed<'_> {
+        Veci8Borrowed::new(self.len, self.data(), self.alpha(), self.offset())
+    }
+
+    pub fn data(&self) -> &[I8] {
+        debug_assert_eq!(self.varlena & 3, 0);
+        debug_assert_eq!(self.kind, VECI8_KIND);
+        unsafe { std::slice::from_raw_parts(self.phantom.as_ptr(), self.len as usize) }
+    }
+
+    pub fn for_borrow(&self) -> Veci8Borrowed<'_> {
+        Veci8Borrowed::new(self.len, self.data(), self.alpha, self.offset)
+    }
+}
+
+impl PartialEq for Veci8Header {
+    fn eq(&self, other: &Self) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        let n = self.len();
+        for i in 0..n {
+            if self.index(i) != other.index(i) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl Eq for Veci8Header {}
+
+impl PartialOrd for Veci8Header {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Veci8Header {
```

Comment:
Do not do so. Implement a util function in `operators_veci8.rs`. Due to unsafe-code-guidelines#256, we want to simplify code in this file so that it's trivial safe.